### DISCUSSION TITLE
Is it possible to use RLS with direct database connections?

### DISCUSSION THREAD

**Discussion by WesleyYue** [supabase/supabase#11056] [Questions] [answered]
I want to write execute raw SQL queries with this setup so that I can use things like json_build_object():

Front-end    ---(pass user auth token)--->    backend    ---(raw sql query with auth token somehow?)--->    pg_pool    --->    postgres

I don't fully understand how the auth.uid() magic works. Is it specifically tied to postgREST? How can I include the auth token manually in a SQL query I create?

**Comment by GaryAustin1** [comment, marked as answer]
You can setup things like PostgREST does doing something like this for basic stuff.
```
set session role authenticated;
set request.jwt.claims to '{"role":"authenticated", "sub":"5950b438-b07c-4012-8190-6ce79e4bd8e5"}';
```
You have to set the role claim if you use auth.roles() at all.  You can set other "jwt" claims besides sub also as needed.
Sub is what auth.uid() extracts.

**Reply by WesleyYue** [reply]
Okay thanks, and how can I authenticate the JWT?

My understanding is:
1. Decode JWT
2. Authenticate access_token to ensure it is valid and authorizes the uid in the jwt **<-- how do I do this step?**
3. Prepend `set session role authenticated; set request.jwt.claims ...` in my raw SQL query.
4. Execute SQL query as normal

**Reply by GaryAustin1** [reply]
I presume this is running from a server.
Most languages have a method, or module to handle jwts.  

Something like jwt.decode (which gets the claims) and jwt.verify which also takes the private key (which you get from your Supabase instance).  Although you can read the claims without verifying, you should also verify with the private key.

If verify does not also check expire, then you check if it is expired from that claim and then just use the claims.

https://supabase.com/docs/learn/auth-deep-dive/auth-deep-dive-jwts

**Reply by WesleyYue** [reply]
Ah I see. Makes sense, thanks.
